lib/gssapi/sanon/sanon_locl.h

```diff
@@ -0,0 +1,70 @@
+/*
+ * Copyright (c) 2019-2020, AuriStor, Inc.
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ *
+ * - Redistributions of source code must retain the above copyright
+ *   notice, this list of conditions and the following disclaimer.
+ *
+ * - Redistributions in binary form must reproduce the above copyright
+ *   notice, this list of conditions and the following disclaimer in
+ *   the documentation and/or other materials provided with the
+ *   distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ * FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ * COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT,
+ * INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
+ * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
+ * STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED
+ * OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ */
+
+#ifndef SANON_LOCL_H
+#define SANON_LOCL_H 1
+
+#include <config.h>
+
+#include <krb5_locl.h>
+#include <gsskrb5_locl.h>
+#include <gssapi.h>
+#include <gssapi_mech.h>
+#include <gssapi_krb5.h>
+
+#include "mech/mech_locl.h"
+
+#include "x25519_ref10.h"
+
+#define SANON_FLAG_LOCAL 1
+#define SANON_FLAG_COMPLETE 2
+
+typedef struct sanon_ctx_desc {
+    uint8_t sk[crypto_scalarmult_curve25519_BYTES];
+    uint8_t pk[crypto_scalarmult_curve25519_BYTES];
+    uint32_t flags;
+    krb5_keyblock session_key;
+    krb5_auth_context_data auth_context; /* need to make krb5 mech happy */
```

Comment:
You might need this to be a `krb5_auth_context` rather than a `krb5_auth_context_data` so you can `krb5_auth_con_free()` it in case the krb5 mech and library allocate more stuff into it.  Have you confirmed there's no leaks?

Reply:
BTW, we now have `gss-token` in-tree.  It might be useful for testing.

Reply:
Only by inspection. SAnon only uses the functionality in `krb5/cfx.c`, which calls:

- `krb5_auth_con_getlocalseqnumber()`
- `krb5_auth_con_setlocalseqnumber()`
- `krb5_auth_con_getlocalsubkey()`
- `krb5_auth_con_getremotesubkey()`
- `krb5_auth_con_getkey()`

none of which allocate anything in `auth_context`.

Reply:
OK, thanks.